File: packages/patterns/src/research-diamond.ts
/**
 * Pattern 01 — the multi-source research diamond.
 *
 * This is a *composition* over {@link diamond}, not a second topology. One
 * scoping node decomposes a question into independent per-source jobs, every
 * source job runs in parallel, and a single fan-in barrier is the only place
 * cross-source comparison happens.
 *
 * What the constructor decides (so a caller cannot make two bundles that differ
 * only by accident):
 *
 * - node identity — `scope`, `source-<key>`, `synthesize`;
 * - source ordering — keys are normalized by Unicode code point by `diamond`;
 * - fan-in — one merge input port per source key, never a positional array;
 * - retry headroom — each source declares `retry.maxAttempts`, which is also
 *   what makes an interrupted source resumable rather than in-doubt;
 * - concurrency and attempt policy derived from the source count.
 *
 * What the constructor does **not** do: it declares no budget, no permission,
 * no network policy and no isolation, because no runtime in this repository
 * enforces any of those. Declaring them here would produce a graph that reads
 * as governed and executes as ungoverned. The bundle manifest states them as
 * documented intent instead — see
 * `examples/patterns/research-diamond/manifest.json`.
 */

import type { JsonSchema, NodeSpec } from "@graph-engineering/core";
import { PatternInputError } from "./errors.js";
import { MAX_PATTERN_ITEMS, safeKey } from "./internal.js";
import { diamond } from "./patterns.js";
import type { KeyedNode, PatternGraph } from "./types.js";

/** One independent, bounded source job. */
export interface ResearchSource {
  /** Stable source key. Names the node suffix and the merge input port. */
  readonly key: string;
  /** The scoped question this source alone is responsible for answering. */
  readonly role: string;
}

export interface ResearchDiamondOptions {
  /** Graph metadata name; must match `^[a-z][a-z0-9-]{0,62}$`. */
  readonly name?: string;
  /** Graph metadata version. Defaults to `1.0.0`. */
  readonly version?: string;
  /** At least one source; keys must be unique. */
  readonly sources: readonly ResearchSource[];
  /**
   * Attempts allowed per source, 1..8. Defaults to 2. A value above 1 is what
   * lets a durable resume re-drive a source whose attempt was interrupted;
   * with 1 an interrupted source settles as failed instead.
   */
  readonly maxAttemptsPerSource?: number;
}

const OBJECT_SCHEMA: JsonSchema = { type: "object" };
const MAX_ATTEMPTS_CEILING = 8;

function sourceNode(source: ResearchSource, maxAttempts: number): NodeSpec {
  return {
    id: `source-${source.key}`,
    kind: "agent",
    inputSchema: OBJECT_SCHEMA,
    outputSchema: OBJECT_SCHEMA,
    config: { sourceKey: source.key, role: source.role },
    retry: { maxAttempts },
    sideEffects: "none",
  };
}

function parseSources(value: unknown): readonly ResearchSource[] {
  if (!Array.isArray(value)) {
    throw new PatternInputError("GE_PATTERN_INVALID_INPUT", "expected an array", "#/sources");
  }
  if (value.length === 0) {
    throw new PatternInputError(
      "GE_PATTERN_EMPTY_COLLECTION",
      "at least one source is required",
      "#/sources",
    );
  }
  if (value.length > MAX_PATTERN_ITEMS) {
    throw new PatternInputError(
      "GE_PATTERN_TOO_MANY_ITEMS",
      `at most ${MAX_PATTERN_ITEMS} sources are allowed`,
      "#/sources",
    );
  }
  const seen = new Set<string>();
  return value.map((item, index) => {
    const path = `#/sources/${index}`;
    if (typeof item !== "object" || item === null || Array.isArray(item)) {
      throw new PatternInputError("GE_PATTERN_INVALID_INPUT", "expected an object", path);
    }
    const entry = item as Record<string, unknown>;
    for (const key of Object.keys(entry)) {
      if (key !== "key" && key !== "role") {
        throw new PatternInputError(
          "GE_PATTERN_UNKNOWN_FIELD",
          `unknown source field '${key}'`,
          `${path}/${key}`,
        );
      }
    }
    const key = safeKey(entry.key, `${path}/key`, "source key");
    // Caught here rather than inside `diamond`, so the reported pointer names
    // the caller's own `sources` array instead of the internal worker list.
    if (seen.has(key)) {
      throw new PatternInputError(
        "GE_PATTERN_DUPLICATE_KEY",
        `duplicate source key '${key}'`,
        `${path}/key`,
      );
    }
    seen.add(key);
    if (typeof entry.role !== "string" || entry.role.length === 0) {
      throw new PatternInputError(
        "GE_PATTERN_INVALID_INPUT",
        "source role must be a non-empty string",
        `${path}/role`,
      );
    }
    return { key, role: entry.role };
  });
}

/**
 * Build the Pattern 01 research diamond.
 *
 * ```text
 *                 ┌── source-a ──┐
 *   scope ────────┼── source-b ──┼──► synthesize
 *                 └── source-c ──┘
 * ```
 *
 * Duplicate source keys, an empty source list, an unknown field and an
 * out-of-range attempt bound are all rejected before a graph exists.
 */
export function researchDiamond(options: ResearchDiamondOptions): PatternGraph {
  if (typeof options !== "object" || options === null || Array.isArray(options)) {
    throw new PatternInputError("GE_PATTERN_INVALID_INPUT", "expected an object", "#");
  }
  const record = options as unknown as Record<string, unknown>;
  for (const key of Object.keys(record)) {
    if (!["name", "version", "sources", "maxAttemptsPerSource"].includes(key)) {
      throw new PatternInputError(
        "GE_PATTERN_UNKNOWN_FIELD",
        `unknown pattern option '${key}'`,
        `#/${key}`,
      );
    }
  }

  const attempts = record.maxAttemptsPerSource ?? 2;
  if (
    typeof attempts !== "number" ||
    !Number.isSafeInteger(attempts) ||
    attempts < 1 ||
    attempts > MAX_ATTEMPTS_CEILING
  ) {
    throw new PatternInputError(
      "GE_PATTERN_INVALID_INPUT",
      `maxAttemptsPerSource must be an integer from 1 through ${MAX_ATTEMPTS_CEILING}`,
      "#/maxAttemptsPerSource",
    );
  }

  const sources = parseSources(record.sources);
  const workers: KeyedNode[] = sources.map((source) => ({
    key: source.key,
    node: sourceNode(source, attempts),
  }));

  return diamond({
    metadata: {
      name: typeof record.name === "string" ? record.name : "research-diamond",
      version: typeof record.version === "string" ? record.version : "1.0.0",
      description: "Decompose one question into independent sources, then merge at one barrier",
    },
    split: {
      id: "scope",
      kind: "transform",
      inputSchema: OBJECT_SCHEMA,
      outputSchema: OBJECT_SCHEMA,
      config: {
        operation: "decompose",
        sources: [...sources.map((source) => source.key)].sort(),
      },
      sideEffects: "none",
    },
    workers,
    merge: {
      id: "synthesize",
      kind: "barrier",
      inputSchema: OBJECT_SCHEMA,
      outputSchema: OBJECT_SCHEMA,
      config: { condition: "all" },
      sideEffects: "none",
    },
    outputKey: "report",
    policies: {
      maxConcurrency: sources.length,
      maxFanOut: sources.length,
      maxDepth: 3,
      // One scope attempt, `attempts` per source, one synthesize attempt.
      maxTotalAttempts: 2 + sources.length * attempts,
    },
  });
}

File: python/src/graph_engineering/patterns/research_diamond.py
"""Pattern 01 — the multi-source research diamond, native Python constructor.

The contract this file must hold is exact: for the same options, the document
returned here must equal the document returned by ``researchDiamond`` in
``@graph-engineering/patterns`` under canonical JSON.  Node identity, source
key normalization, edge identity, merge port names, the reserved pattern
labels, the named output key and the derived policies are all decided here in
the same way the TypeScript constructor decides them.

Like its TypeScript peer this constructor declares no budget, no permission,
no network policy and no isolation.  No runtime in this repository enforces
any of those, and emitting them into the graph would make an ungoverned run
read as a governed one.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Final

from ..models import JsonValue

_KEY_IDENTIFIER: Final = re.compile(r"^[A-Za-z][A-Za-z0-9_.-]{0,63}$")
_GRAPH_NAME: Final = re.compile(r"^[a-z][a-z0-9-]{0,62}$")
_RESERVED_KEYS: Final = frozenset({"__proto__", "constructor", "prototype"})
_MAX_SOURCES: Final = 100
_MAX_ATTEMPTS_CEILING: Final = 8

_PATTERN_LABEL: Final = "graphengineering.reacher-z.github.io/pattern"
_CAPABILITY_LABEL: Final = "graphengineering.reacher-z.github.io/runtime-capability"
_API_VERSION: Final = "graphengineering.reacher-z.github.io/v1alpha1"

_OBJECT_SCHEMA: Final[dict[str, JsonValue]] = {"type": "object"}


class PatternInputError(ValueError):
    """A rejected constructor input, carrying a stable code and JSON pointer.

    The codes are the same strings the TypeScript ``PatternInputError`` uses,
    so a cross-language test can compare refusals and not just successes.
    """

    def __init__(self, code: str, message: str, path: str) -> None:
        super().__init__(f"{message} at {path}")
        self.code = code
        self.path = path


@dataclass(frozen=True, slots=True)
class ResearchSource:
    """One independent, bounded source job."""

    key: str
    role: str


def _object_schema() -> dict[str, JsonValue]:
    return dict(_OBJECT_SCHEMA)


def _safe_key(value: object, path: str, description: str) -> str:
    if (
        not isinstance(value, str)
        or _KEY_IDENTIFIER.match(value) is None
        or value in _RESERVED_KEYS
    ):
        raise PatternInputError(
            "GE_PATTERN_INVALID_IDENTIFIER",
            f"{description} must match {_KEY_IDENTIFIER.pattern} and not be a reserved key",
            path,
        )
    return value


def _normalized_sources(sources: object) -> tuple[ResearchSource, ...]:
    if isinstance(sources, str) or not isinstance(sources, (list, tuple)):
        raise PatternInputError("GE_PATTERN_INVALID_INPUT", "expected a sequence", "#/sources")
    if len(sources) == 0:
        raise PatternInputError(
            "GE_PATTERN_EMPTY_COLLECTION", "at least one source is required", "#/sources"
        )
    if len(sources) > _MAX_SOURCES:
        raise PatternInputError(
            "GE_PATTERN_TOO_MANY_ITEMS",
            f"at most {_MAX_SOURCES} sources are allowed",
            "#/sources",
        )

    parsed: list[ResearchSource] = []
    seen: set[str] = set()
    for index, item in enumerate(sources):
        path = f"#/sources/{index}"
        if isinstance(item, ResearchSource):
            key_value: object = item.key
            role_value: object = item.role
        elif isinstance(item, dict):
            unknown = sorted(set(item) - {"key", "role"})
            if unknown:
                raise PatternInputError(
                    "GE_PATTERN_UNKNOWN_FIELD",
                    f"unknown source field '{unknown[0]}'",
                    f"{path}/{unknown[0]}",
                )
            key_value = item.get("key")
            role_value = item.get("role")
        else:
            raise PatternInputError(
                "GE_PATTERN_INVALID_INPUT", "expected a mapping or ResearchSource", path
            )

        key = _safe_key(key_value, f"{path}/key", "source key")
        if key in seen:
            raise PatternInputError(
                "GE_PATTERN_DUPLICATE_KEY", f"duplicate source key '{key}'", f"{path}/key"
            )
        seen.add(key)
        if not isinstance(role_value, str) or role_value == "":
            raise PatternInputError(
                "GE_PATTERN_INVALID_INPUT",
                "source role must be a non-empty string",
                f"{path}/role",
            )
        parsed.append(ResearchSource(key=key, role=role_value))

    # Unicode code-point order, exactly as the TypeScript ``diamond`` normalizes
    # keyed workers. Python string comparison is already code-point ordered.
    return tuple(sorted(parsed, key=lambda source: source.key))


def _source_node(source: ResearchSource, max_attempts: int) -> dict[str, JsonValue]:
    return {
        "config": {"role": source.role, "sourceKey": source.key},
        "id": f"source-{source.key}",
        "inputSchema": _object_schema(),
        "kind": "agent",
        "outputSchema": _object_schema(),
        "retry": {"maxAttempts": max_attempts},
        "sideEffects": "none",
    }


def research_diamond(
    *,
    sources: object,
    name: str = "research-diamond",
    version: str = "1.0.0",
    max_attempts_per_source: int = 2,
) -> dict[str, JsonValue]:
    """Build the Pattern 01 research diamond as portable Graph IR.

    ``scope`` fans out to one ``source-<key>`` node per source, and every
    source result reaches a unique ``synthesize`` input port named by its key.
    ``synthesize`` is the single fan-in barrier and the named graph output.

    The returned value is a plain portable-JSON document.  It is not compiled
    here; callers pass it to :func:`graph_engineering.compile_graph`, which is
    the only authority on graph validity.
    """

    if not isinstance(name, str) or _GRAPH_NAME.match(name) is None:
        raise PatternInputError(
            "GE_PATTERN_INVALID_IDENTIFIER",
            f"graph metadata name must match {_GRAPH_NAME.pattern}",
            "#/name",
        )
    if not isinstance(version, str) or version == "":
        raise PatternInputError(
            "GE_PATTERN_INVALID_INPUT",
            "graph metadata version must be a non-empty string",
            "#/version",
        )
    if (
        type(max_attempts_per_source) is not int
        or max_attempts_per_source < 1
        or max_attempts_per_source > _MAX_ATTEMPTS_CEILING
    ):
        raise PatternInputError(
            "GE_PATTERN_INVALID_INPUT",
            f"max_attempts_per_source must be an integer from 1 through {_MAX_ATTEMPTS_CEILING}",
            "#/maxAttemptsPerSource",
        )

    normalized = _normalized_sources(sources)
    keys = [source.key for source in normalized]
    # Built element by element rather than annotated from `sorted`: `list` is
    # invariant, so `list[str]` is not assignable to `list[JsonValue]`.
    declared_keys: list[JsonValue] = list(sorted(keys))

    nodes: list[JsonValue] = [
        {
            "config": {"operation": "decompose", "sources": declared_keys},
            "id": "scope",
            "inputSchema": _object_schema(),
            "kind": "transform",
            "outputSchema": _object_schema(),
            "sideEffects": "none",
        }
    ]
    nodes.extend(_source_node(source, max_attempts_per_source) for source in normalized)
    nodes.append(
        {
            "config": {"condition": "all"},
            "id": "synthesize",
            "inputSchema": _object_schema(),
            "kind": "barrier",
            "outputSchema": _object_schema(),
            "sideEffects": "none",
        }
    )

    edges: list[JsonValue] = []
    index = 1
    for key in keys:
        edges.append(
            {
                "id": f"ge-diamond-{index:04d}",
                "from": {"node": "scope"},
                "to": {"node": f"source-{key}"},
                "mode": "value",
            }
        )
        index += 1
    for key in keys:
        edges.append(
            {
                "id": f"ge-diamond-{index:04d}",
                "from": {"node": f"source-{key}"},
                "to": {"node": "synthesize", "port": key},
                "mode": "value",
            }
        )
        index += 1

    return {
        "apiVersion": _API_VERSION,
        "kind": "Graph",
        "metadata": {
            "description": (
                "Decompose one question into independent sources, then merge at one barrier"
            ),
            "labels": {
                _PATTERN_LABEL: "diamond/v1alpha1",
                _CAPABILITY_LABEL: "dag/v1alpha1",
            },
            "name": name,
            "version": version,
        },
        "inputSchema": _object_schema(),
        "outputSchema": _object_schema(),
        "entrypoints": ["scope"],
        "outputs": {"report": {"node": "synthesize"}},
        "nodes": nodes,
        "edges": edges,
        "policies": {
            "maxConcurrency": len(normalized),
            "maxDepth": 3,
            "maxFanOut": len(normalized),
            "maxTotalAttempts": 2 + len(normalized) * max_attempts_per_source,
        },
    }
